### PR TITLE
Add required fields proposal

### PR DESCRIPTION
**Pull request by mpvl**
This is a complete rewrite and redo of the original require fields proposal.

### CODE REVIEW

File: designs/1951-required-fields-v2.md [hunk 1]
```diff
@@ -0,0 +1,352 @@
+
+
+
+This proposal replaces https://github.com/cue-lang/cue/issues/822.
+Thanks to all of the comments and suggestions contributed there.
+The proposal presented here is considerably simpler and changes less
+about the current semantics.
+
+# Objective
+We introduce a new field type: required fields, with the aim to address
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: Singular and plural don't work well together here.
```suggestion
We introduce a new field type, a _required_ field, with the aim to address
```

File: designs/1951-required-fields-v2.md [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+various shortcomings of CUE while at the same time allowing for simpler semantics for other features we wish to introduce in CUE.
+
+# Background
+One of CUE’s strengths is that it allows specifying types, constraints,
+and values within a unified framework. This means that fields in CUE have many different roles.
+For example, they can define simple JSON-like data, “templatize” or add derived fields that
+compute fields from these data fields, add constraints to data, or define top-level schema types (also constraints).
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: Given that "templatize" isn't a conventional word, how about making it subordinate to the more English explanation?

```suggestion
For example, they can define simple JSON-like data, add derived fields that
compute fields from these data fields (“templatize”), add constraints to data, or define top-level schema types (also constraints).
```

File: designs/1951-required-fields-v2.md [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+
+## Output and interpretation modes
+There are two major modes for outputting and interpreting CUE: 
+schema: output a schema with all constraints intact
+data: outputs only concrete data after validating all constraints
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: This doesn't read well in markdown (I think it wants to be a bulleted list). "output" feels a bit awkward as an active verb. Maybe "produce" or "generate" might read better?

```suggestion
There are two major modes for outputting and interpreting CUE: 
- schema: produce a schema with all constraints intact
- data: produce only concrete data after validating all constraints
```

File: designs/1951-required-fields-v2.md [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+
+The schema mode corresponds to the `cue def` command.
+The data mode corresponds to `cue export`. `cue eval` lives in a limbo land in between the two
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
The data mode corresponds to `cue export`. `cue eval` lives at a poorly specified point in between the two
```
?

File: designs/1951-required-fields-v2.md [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+and should probably be deprecated or could at least benefit from some clarifications.
+This is covered in https://github.com/cue-lang/cue/issues/337.
+
+### Schema mode
+The main goal of evaluating in schema mode are:
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
The main goals of schema-mode evaluation are:
```
?

File: designs/1951-required-fields-v2.md [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+
+1. Simplification of schema, for instance after unifying two schemas.
+1. Validate that a schema, or the unification of two schemas,
+2. is internally consistent, that is, has no errors.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: You don't want three items here.

```suggestion
1. Simplification of schema (after unifying two schemas, for instance)
1. Validate that a schema, or the unification of two schemas, is internally consistent; that is, there are no errors.
```

File: designs/1951-required-fields-v2.md [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+Schema mode preserves all constraints: outputting CUE in schema mode requires all fields,
+including optional ones, to be output.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: Again, "output" seems clunky here.

```suggestion
Schema mode preserves all constraints: generating CUE in schema mode requires all fields,
including optional ones, to be produced. 
```

File: designs/1951-required-fields-v2.md [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+### Data mode
+In data mode, all [regular fields](https://cuelang.org/docs/references/spec/#definitions-and-hidden-fields)
+must be concrete, recursively, and are output.
+It is an error for values to be non-concrete, like `a: string`,
+or yield an incomplete error, like `a: b + c`, where `b` or `c` are not concrete.
+
+### Validation
+Validation can occur in either mode.
+
+`cue vet` is essentially used in both modes of interpretation.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: "essentially" is redundant here.

```suggestion
`cue vet` is used for both modes of interpretation.
```

File: designs/1951-required-fields-v2.md [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+By default it uses schema mode, but it can be explicitly set to a data mode by using the `-c` flag.
+
+## Field Semantics
+We identified the following list of common interpretation or use cases for field semantics:
+
+1. A field is defined with the given value, which must evaluate without error (regular field)
+1. A field _may_ be defined, in which case given constraints apply (optional field).
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: "given" reads awkwardly here IMHO, and we could probably make "constraints" singular, given that we can always consider it as a single constraint.

```suggestion
1. A field _may_ be defined, in which case the associated constraint applies (optional field).
```
?

File: designs/1951-required-fields-v2.md [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+1. A field _must_ be defined elsewhere (required field).
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
1. A field _must_ be defined elsewhere and satisfy an associated constraint (required field).
```

File: designs/1951-required-fields-v2.md [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+1. A field must not exist (optional field constraint where the constraint is an error value).
+1. A field is only defined if some condition is met (conditional field).
+
+In effect there are two types of fields: those that provide constraints and those that populate values. Below we discuss these use cases within particular usage contexts.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md:59]: It would be good to briefly say which of the above items fall into which categories (maybe in parentheses after "those that provide constraints" and "those that populate values").

I'm not sure myself which category the latter item falls into.

File: designs/1951-required-fields-v2.md [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+
+### Data and export
+In CUE, we classify data as [regular fields](https://cuelang.org/docs/references/spec/#definitions-and-hidden-fields)
+(so no definitions or hidden fields), with only concrete values.
+In a nutshell, any JSON value is data in CUE.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
Any JSON value is data in CUE.
```

File: designs/1951-required-fields-v2.md [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+
+In contrast to JSON, but like YAML, CUE permits fields with top-level references to other data fields
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: What does "top-level" mean here?

File: designs/1951-required-fields-v2.md [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+1. If the user specifies a field `x`, apply constraint `c`.
+1. The user _must_ specify a field `x` with constraint `c`.
+
+Orthogonally to this, the user may wish to specify that `c` must be concrete.
+This is not done explicitly, but rather is enforced implicitly in data mode
+by requiring that any regular field (so not a constraint `a?: x`) must be concrete.
+
+Case 1 can be expressed as:
+```
+x? : c
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: Personally, I prefer the colon to nestle up to the LHS (`cue fmt` and convention JSON formatting concurs)

```suggestion
x?: c
```

File: designs/1951-required-fields-v2.md [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+```
+Case 2 is currently expressed as:
+```
+x : c
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
x: c
```

File: designs/1951-required-fields-v2.md [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+```
+The requirement that the user must specify a field is only implicit.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
The requirement that the user must specify a field is implicit.
```

File: designs/1951-required-fields-v2.md [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+This works as desired when `c` is not concrete: because “non-optional” fields are required to be concrete,
+it will force the user to specify the field with the concrete value.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
This works as desired when `c` is not concrete: because “non-optional” fields are required to be concrete.
It will force the user to specify the field with the concrete value.
```

Also, do we actually need the scare quotes around "non-optional" ?

File: designs/1951-required-fields-v2.md [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+
+It is not possible, however, to force a user to specify a field with a specific concrete value.
+This includes any list or struct.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
It is not possible, however, to force a user to specify a field with a specific concrete value,
which includes any list or struct.
```

File: designs/1951-required-fields-v2.md [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+This is because there is no syntactic distinction between a field that is a constraint,
+but required, and a regular field: the required property is implied by the fact that regular
+fields must be concrete upon output.
+That is, a regular field containing a concrete value that must be specified is
+indistinguishable from a regular field with a concrete value specified by the user.
+
+See, for instance, https://github.com/cue-lang/cue/issues/740 where this is problematic.
+
+### Derived fields and templating
+The purpose of derived fields and templating is to automatically populate fields for a user.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
The purpose of templating, including deriving fields, is to automatically populate fields for a user.
```
But this is a very didactic (and potentially controversial statement), because someone else might consider that deriving a field to be schema-like in some cases.

How about saying something like "We consider that the purpose of templating..." or similar to make it a bit softer?

File: designs/1951-required-fields-v2.md [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+Possible use cases are:
+
+* providing a set of defaults, 
+* defining a higher level of abstraction that is mapped to some more concrete representation, or
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
* defining a higher level of abstraction that is mapped to some more concrete representation
```

File: designs/1951-required-fields-v2.md [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+Possible use cases are:
+
+* providing a set of defaults, 
+* defining a higher level of abstraction that is mapped to some more concrete representation, or 
+* converting between two versions of an API.
+
+Examples of such fields are:
+
+1. Defaults: `a: *1 | int`
+1. Derived values: `a: b + c`
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: We've called this "derived fields" before, so maybe keep to that terminology?
```suggestion
1. Derived fields: `a: b + c`
```

File: designs/1951-required-fields-v2.md [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+1. Derived constraints: `a: b + c`, but only if `b + c` is defined.
+
+All derived fields use regular, non-optional, fields.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: This is an assertion that's not necessarily true (it's perfectly feasible to define a required or optional field as a derived field AFAICS). Maybe "should use" ? I'm not even sure that that's true tbh.

File: designs/1951-required-fields-v2.md [hunk 23]
```diff
@@ -0,0 +1,352 @@
...
+If a field should only be added conditionally, one currently has to revert to rather verbose methods, such as
+```
+a?: int
+b?: int
+if a != _|_ && b != _|_ {
+  c: a + b
+}
+```
+Case 3 is common (see https://github.com/cue-lang/cue/issues/1232).
+Having something much more compact would be nice and some thoughts for a proposal for that are in progress.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: I found this hard to parse.
```suggestion
Having something much more compact would be nice: we have some thoughts for a possible proposal to address this.
```
?

File: designs/1951-required-fields-v2.md [hunk 24]
```diff
@@ -0,0 +1,352 @@
...
+#DefPolicy: #Def & {
+    name?:  strings.MinRunes(1)
+    speed?: >100
+}
+```
+# Proposal
+CUE currently has the “optional field constraint”, denoted `foo?: value`.
+
+We propose adding a “required field constraint”, denoted `foo!: value`, which is like `foo?: value`,
+but requires a regular value be provided for `foo` (a field `foo:` without  `!:` or ?:`).
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
but requires a regular value be provided for `foo` (a field `foo:` without  `!:` or `?:`).
```

File: designs/1951-required-fields-v2.md [hunk 25]
```diff
@@ -0,0 +1,352 @@
...
+
+We refer to optional field constraints and required field constraints collectively as “field constraints”.
+
+As a general rule, all data and data templating would be defined with regular fields,
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
As a general rule, we consider that all data and data templating should be defined with regular fields,
```

File: designs/1951-required-fields-v2.md [hunk 26]
```diff
@@ -0,0 +1,352 @@
...
+whereas schemas would be defined in terms of field constraints.
+Of course, CUE being CUE, mixing these two fields is allowed:
+this rule is not a restriction but suggested as a matter of style and proper usage.
+
+## Semantics
+Informally, requiredness `!` is a property of a field that is unified independently of a field value.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]: I don't think the `!` helps much here.
```suggestion
Informally, requiredness is a property of a field that is unified independently of a field value.
```

File: designs/1951-required-fields-v2.md [hunk 27]
```diff
@@ -0,0 +1,352 @@
...
+{foo!: 3} & {foo: <=4}	   	→  {foo: 3}  // X
+
+{foo!: int} & {foo: int} 	  	→  {foo: int}
+{foo!: int} & {foo?: <1}   	→  {foo!: <1}
+{foo!: int} & {foo: <=3}   	→  {foo: <=3}
+{foo!: int} & {foo: 3}  		→  {foo: 3}
+```
+Note the subtle, but important difference between this proposal and #822: `foo!: value`
+only requires that a field be specified; it does not require it to be concrete.
+As we are no longer proposing a change to the semantics of `x: string`, though, the net effect is roughly the same.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md]:
```suggestion
However as we are no longer proposing a change to the semantics of `x: string`, the net effect is roughly the same.
```

File: designs/1951-required-fields-v2.md [hunk 28]
```diff
@@ -0,0 +1,352 @@
...
+
+As a consequence, though, it is now possible to create a concrete value without
+having the user actually provide it (see the line marked `X` in the above example).
+So this loses the strictness of the original proposal.
+In practice, this likely is not much of an impediment: JSON and YAML cannot specify such fields,
+and thus this would not limit the ability to validate such files: they still need to specify
+the actual concrete value to pass.
```
Comment (rogpeppe) [inline designs/1951-required-fields-v2.md:186]: It might be good to say that the proposal is technically complete at this point?

```suggestion
the actual concrete value to pass.

That completes the proposal.
```

**Review by rogpeppe — COMMENTED** [review]
Good stuff! Lots of grammar nits and nothing more. This review gets as far as the Examples section and no further.

File: designs/1951-required-fields-v2.md (hunk 13 above)
Comment (mpvl) [reply designs/1951-required-fields-v2.md]: Good point. Don't quite remember. I think it can be dropped.

File: designs/1951-required-fields-v2.md (hunk 22 above)
Comment (mpvl) [reply designs/1951-required-fields-v2.md]: Good point. That is absolutely true.

I'll removed that sentence.
